cns/client/client.go

```diff
@@ -1009,3 +1010,87 @@ func (c *Client) GetHomeAz(ctx context.Context) (*cns.GetHomeAzResponse, error)
 
 	return &getHomeAzResponse, nil
 }
+
+// GetEndpoint calls the EndpointHandlerAPI in CNS to retrive the state of a given EndpointID
+func (c *Client) GetEndpoint(ctx context.Context, endpointID string) (*restserver.GetEndpointResponse, error) {
+	getEndpoint := cns.EndpointRequest{
+		EndpointID: endpointID,
+	}
+	var body bytes.Buffer
+
+	if err := json.NewEncoder(&body).Encode(getEndpoint); err != nil {
+		return nil, errors.Wrap(err, "failed to encode getEndpoint")
+	}
+
+	u := c.routes[cns.EndpointsPath]
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, u.String(), &body)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to build request")
+	}
+	req.Header.Set(headerContentType, contentTypeJSON)
+	res, err := c.client.Do(req)
+	if err != nil {
+		return nil, errors.Wrap(err, "http request failed")
+	}
+
+	defer res.Body.Close()
+
+	if res.StatusCode != http.StatusOK {
+		return nil, errors.Errorf("http response %d", res.StatusCode)
+	}
+
+	var response restserver.GetEndpointResponse
+	err = json.NewDecoder(res.Body).Decode(&response)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to decode IPConfigResponse")
+	}
+
+	if response.Response.ReturnCode != 0 {
+		return nil, errors.New(response.Response.Message)
+	}
+
+	return &response, nil
+}
+
+// UpdateEndpoint calls the EndpointHandlerAPI in CNS to update the state of a given EndpointID with either HNSEndpointID or HostVethName
+func (c *Client) UpdateEndpoint(ctx context.Context, endpointID, hnsID, vethName string) (*cns.UpdateEndpointResponse, error) {
+	updateEndpoint := cns.EndpointRequest{
+		EndpointID:    endpointID,
+		HnsEndpointID: hnsID,
+		HostVethName:  vethName,
+	}
+	var body bytes.Buffer
+
+	if err := json.NewEncoder(&body).Encode(updateEndpoint); err != nil {
+		return nil, errors.Wrap(err, "failed to encode updateEndpoint")
+	}
+
+	u := c.routes[cns.EndpointsPath]
+	req, err := http.NewRequestWithContext(ctx, http.MethodPatch, u.String(), &body)
```

Comment:
PATCH is not the right verb here. [[MDN]](https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/PATCH):
> A PATCH request is considered a set of instructions on how to modify a resource. Contrast this with [PUT](https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/PUT); which is a complete representation of a resource.  

This is a full representation of the resource, so it's either PUT or POST, depending on exactly what the desired effect is (wrt idempotency):
> The difference between PUT and [POST](https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/POST) is that **_PUT is idempotent_**: calling it once or several times successively has the same effect (that is no side effect), whereas successive identical [POST](https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/POST) requests may have additional effects, akin to placing an order several times.

Here, I think you want to PUT.

Reply:
Additionally, building on the RESTful semantics context I previously mentioned, PUTing a resource in this collection should probably be done at:
`PUT /network/endpoints/{{endpointID}}`

with the non-"uniquely identifying" content of the object in the Body.

Reply:
@rbtr Sorry I missed this comment. I'm wondering what is the resource in here? Is it the the full state which has:
EndpointID, POdName, POdNamespace, IPAddress, VethName, HNSID .... or the individual attributes?
In this API we are only assigning values to two attributes within the State. 
So the question is that is this consider a modification of state resource (PATCH) or changing the complete representation of HNSID (a PUT) ?